include/class_loader/class_loader.hpp

```diff
@@ -124,10 +124,17 @@ class ClassLoader
   template<class Base>
   std::shared_ptr<Base> createInstance(const std::string & derived_class_name)
   {
-    return std::shared_ptr<Base>(
-      createRawInstance<Base>(derived_class_name, true),
-      std::bind(&ClassLoader::onPluginDeletion<Base>, this, std::placeholders::_1)
-    );
+    try {
+      return std::shared_ptr<Base>(
+        createRawInstance<Base>(derived_class_name, true),
+        std::bind(&ClassLoader::onPluginDeletion<Base>, shared_from_this(), std::placeholders::_1)
+      );
+    } catch (std::bad_weak_ptr & e) {  // This is not a shared_ptr
+      return std::shared_ptr<Base>(
+        createRawInstance<Base>(derived_class_name, true),
+        std::bind(&ClassLoader::onPluginDeletion<Base>, this, std::placeholders::_1)
+      );
```

Comment:
@ahcorde the not so great aspect of this is that client code may run into it (and leak) inadvertently, as opposed to doing so on purpose. Thinking out loud, maybe we should document that `ClassLoader::enabled_shared_from_this()` is expected to be valid and add a deprecation warning when it's not. Eventually we can get rid of the non-owning case. WDYT @clalancette ?

Reply:
@clalancette ?

Reply:
friendly ping @clalancette

Reply:
I stand by my suggestion if we want to move forward.